### PR TITLE
Refactor meta data ingestion

### PR DESCRIPTION
**Pull request by profxj**
Refactor the definition and ingestion of meta data in PypeIt.
The primary philosophies guiding the refactor are:

1. Generate a data model for the meta data for all spectrographs to follow, which includes a set of required meta data to run PypeIt
1. Push most of the metadata ingestion details onto the spectrograph classes
1. Enable meta data parsing from an input file (via the Spectrograph class)

This new doc provides a bit more detail:

https://github.com/pypeit/PypeIt/blob/refactor_meta_loading/doc/metadata.rst

I don't think there are any fundamental changes to `PypeItMetaData` aside
from a revamped `_build()` method.

Below is the more detailed list of CHANGES.  
Passing all tests (on Cooked v0.94) and running fine so far on the dev suite but
requires the `refactor_meta` branch which I will submit a PR on shortly.

@jhennawi -- Note that a Setup block is now required in the PypeIt file to run.
I have modified the pypeit files in the Dev suite accordingly.

CHANGES:

- Nearly resurrected simple_calib
- New method to build the fitstbl of meta data
- Refactor handling of meta data including a data model defining core and additional meta data
- Replaces metadata_keys with pypeit_file_keys for output to PypeIt file
- Updates new metadata approach for VLT, Keck, Lick, Gemini instruments
- Remove PypeItSetup call from within PypeIt
- Remove lacosmic specific method in Spectrograph;  replaced with config_specific_par
- setup block now required when running on a PypeIt file

### CODE REVIEW

**Comment by jhennawi** [comment]
Hi @profxj, 

This looks great and long overdue. A few comments: 

-- Is now the time to get rid of the number = 0 defaults that get written to the pypeit files? I think just having the block with 

[rdx]
[wavelengths]
 etc. 

would suffice without the number=X stuff. The number of required files should just be set in the defaults in the spectrograph class. 

-- Can you elaborate the new setup block is required in the pypeit file. I looked at them in your branch of the dev suite and it seemed redundant to me, i.e. you are reading setup A and then every column has setup = A. Why are you requiring the setup block in these cases? It seems to just add a failure mode without adding any functionality since that information is in the PypeIt file already. There were other cases like DEIMOS where the individual detectors are listed in the setup block of PypeIt file, but this does not seem necessary since meta-data should not ever change for different detectors. 

Thanks,
Joe

**Comment by profxj** [comment]
The idea is to replace having the setup column with 
this setup block (which also describes the setup being
reduced).  We probably should minimize the detector info
for cases like DEIMOS.

**Comment by jhennawi** [comment]
How does the code use this setup information anyway? I mean for most of the examples it is simply 'A'?

Since the pypeit file only makes sense for a single setup, I agree that it makes more sense to have this block rather than list the same thing many times in a column. However, I guess the details of the setup need to be somewhere for each file, i.e. we need that as a log file.

**Comment by kbwestfall** [comment]
A few initial comments before I keep going (sorry I didn't do this sooner):
 - I can't update the docs because not all the instruments have
   init_meta properly defined.  I either need to change the script that
   generates the par docs, or we need to implement this function for all
   spectrographs.  I'd rather we do the latter because we'll need to do
   that anyway.  I can help with this.

 - I feel like we're flirting with a circular import.  I understand the
   reasoning, but I think having Spectrograph depend on functions
   defined in metadata confuses the code tree structure.  I.e., what's
   our lowest level dependency, metadata or Spectrograph?  We get away
   with the current structure because we never actually have to import
   the spectrograph into metadata, we just use it.  But I worry about
   long-term viability.

 - I like that the _build function in metadata is much simpler now!

 - I also like that we've removed the call to PypeItSetup within PypeIt.

 - I'd prefer that we keep the setup column because it allows the user
   to by-hand change the setup assigned to each frame.  But maybe this
   is a functionality we don't need?

File: pypeit/spectrographs/keck_lris.py [hunk 1]
```diff
@@ -104,10 +105,67 @@ def header_keys(self):
...
+            self.meta: dict (generated in place)
+
+        """
+        meta = {}
+        # Required (core)
+        meta['ra'] = dict(ext=0, card='RA')
+        meta['dec'] = dict(ext=0, card='DEC')
+        meta['target'] = dict(ext=0, card='TARGNAME')
+        meta['decker'] = dict(ext=0, card='SLITNAME')
+        meta['binning'] = dict(card=None, compound=True)
```
Comment (kbwestfall) [inline pypeit/spectrographs/keck_lris.py]: Do we ever use the `compound` component of meta?  My rigid C brain is freaking out about some keys in these meta dictionaries not being defined in *all* dictionaries...

Comment (kbwestfall) [reply pypeit/spectrographs/keck_lris.py]: Okay, found where it's used.  My C brain is still unhappy, but I'll deal...

**Comment by kbwestfall** [comment]
Adding some init_meta methods now...

File: pypeit/scienceimage.py [hunk 2]
```diff
@@ -702,14 +702,14 @@ def build_crmask(self, stack, ivar=None):
         varframe = utils.calc_ivar(ivar)
         saturation = self.spectrograph.detector[self.det-1]['saturation']
         nonlinear = self.spectrograph.detector[self.det-1]['nonlinear']
-        sigclip, objlim = self.spectrograph.get_lacosmics_par(proc_par,binning=self.binning)
+        #sigclip, objlim = self.spectrograph.get_lacosmics_par(proc_par,binning=self.binning)
```
Comment (jhennawi) [inline pypeit/scienceimage.py]: Hi @profxj can you explain how you now deal with this functionality using your new meta data construct? It seems like you need to dynamically update the parset depending on the file in question that you are dealing with in order to make these binning dependent adjustments to the CR finding parameters. But perhaps I'm being ignorant here, and missed something simple. In my opinion parsets should be immutable constructs and not change depending on which file is read in.

**Comment by kbwestfall** [comment]
After trying to implement the init_meta method in all the spectrographs, I realized that the additional keys used by *any* instrument are defined in a single place.  Was there a reason to do it this way?  I.e., why is  `define_additional_meta` defined in PypeItMetaData and not in Spectrograph so that each spectrograph can overwrite it?

**Comment by profxj** [comment]
The top-level goal was to generate a (ideally rigid) data model for the 
meta-data used throughout `PypeIt`.  So while each instrument might have
a way for recording the exposure time, you can be certain in `PypeIt` that
it will be a `float` and in seconds, etc.  These are all defined in `core_meta`
and `additional_meta` (not the best name).  

As to whether to define those in `PypeItMetaData` or `Spectrograph`, 
I am somewhat agnostic but they are carried around by `PypeItMetaData`
and accessed via `PypeItMetaData`.  They are only generated (and occasionally
used) by `Spectrograph`.

File: pypeit/core/save.py [hunk 3]
```diff
@@ -351,14 +351,19 @@ def save_1d_spectra_hdf5(slf, fitsdict, clobber=True):
...
+                         outfile, helio_dict=None, telescope=None, overwrite=True,
                          update_det=None):
     """ Write 1D spectra to a multi-extension FITS file
 
     Parameters
     ----------
     specobjs : SpecObjs object
     header : dict or Row (dict-like)
+    pypeline : str
+      Name of instrument
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/save.py]: fixed

File: pypeit/core/arc.py [hunk 4]
```diff
@@ -1035,6 +1034,9 @@ def simple_calib(msarc, aparm, censpec, nfitpix=5, get_poly=False,
                 idsion[jj] = llist['Ion'][imnw]
                 msgs.info("Identifying arc line: {:s} {:g}".format(idsion[jj],ids[jj]))
     else:
+        # DEPRECATED!!
+        debugger.set_trace()
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/arc.py]: I will clean this one and the next one in a future PR.
No one uses this method at the moment.

File: pypeit/core/arc.py [hunk 5]
```diff
@@ -1083,10 +1085,14 @@ def simple_calib(msarc, aparm, censpec, nfitpix=5, get_poly=False,
                 debugger.plot1d(yprep)
             else:
                 msgs.error('Insufficient lines to auto-fit.')
+        '''
 
     # Debug
     msgs.work('Cross correlate here?')
 
+    debugger.set_trace() # STOPPED HERE
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/arc.py]: see above

File: pypeit/core/save.py [hunk 6]
```diff
@@ -385,7 +390,8 @@ def save_1d_spectra_fits(specObjs, header, pypeline, outfile, helio_dict=None, t
         try:
             prihdu.header['MJD-OBS'] = header['MJD-OBS']
         except KeyError:
-            prihdu.header['MJD-OBS'] = header['time']  # recorded as 'time' in fitstbl
+            prihdu.header['MJD-OBS'] = header['mjd']  # recorded as 'mjd' in fitstbl
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/core/save.py]: I've swapped the ordering.  I think this is ok.
It is only an issue when fluxing (so far) with 
a script distinct from `run_pypeit`

File: pypeit/par/util.py [hunk 7]
```diff
@@ -519,6 +521,7 @@ def parse_pypeit_file(ifile, file_check=True):
         data_files, frametype, usrtbl = _read_data_file_table(lines[s:e], file_check=file_check)
     is_config[s-1:e+1] = False
     if len(data_files) == 0 and file_check:
+        import pdb; pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypeit/par/util.py]: removed

File: pypeit/spectrographs/keck_lris.py (hunk 1 above)
Comment (profxj) [reply pypeit/spectrographs/keck_lris.py]: Fair point.  In hindsight, having a method that builds each of these
(and all the default keys) would be good practice.  e.g.

```
def mdict(ext=0, compound=False, etc...):
    new_dict = {}
    new_dict['ext'] = ext
    etc.
    return new_dict
```

I could be convinced to do this if so commanded.  ;)

File: pypeit/scienceimage.py (hunk 2 above)
Comment (profxj) [reply pypeit/scienceimage.py]: Yes, there are now two calls using the `Spectrograph` to set PypeItPar:

- The call using the output from `default_pypeit_par()`
- A call using `config_specific_par` which takes in a science frame filename

Both are executed at the start of `PypeIt`.  Perhaps ideally they'd be done
in the same call.  Looking at it again, I think I could do that.  Doing so now..

But to finish the response to your question, `PypeItPar` remains immutable after
the initial call(s) to `Spectrograph`.  The advance here is that parameters can be
changed at the onset depending on the instrument configuration.

Comment (jhennawi) [reply pypeit/scienceimage.py]: Hi @profxj, 

Okay but as far as I can see binning is still its separate category not included in this parset. For X-shooter we have 1x1 binned arcs applied to data of arbitrary binning which is why I ask.

**Comment by profxj** [comment]
Passing all standard tests and I believe the Dev suite too (VLT is failing
after metadata, I think because a linelist is missing).

I think we are ready to merge.

**Comment by jhennawi** [comment]
Hi @kbwestfall and @profxj, 

Can you please take this opportunity to get rid of this number = XX that it is in the pypeit files. It serves no purpose, causes confusion, and adds failure modes. 

Thanks,
Joe

**Comment by jhennawi** [comment]
Hi @kbwestfall and @profxj,

There appears to be an inconsistency with the setup block that you are writing out and now requiring in the pypeit file. For example: 

If I run: 

pypeit_setup -r .RAW_DATA/VLT_XSHOOTER/NIR/ -s vlt_xshooter_nir -b

I get a setup block in my pypeit file that looks like this: 

# Setup
setup read
 A:
   --:
     dichroic: none
     disperser: {angle: none, name: default}
     slit: {decker: 1.0x12.6, slitlen: none, slitwid: none}
   '01': {binning: '1,1', det: 1, namp: 1}
setup end

But when I try to run pypeit on such files I get the error: 

[ERROR]   :: util.py 546 parse_pypeit_file() - Need to add setup info to your PypeIt file in the setup block!

If I change the setup block to: 

# Setup
setup read
 Setup A
setup end

Things run just fine. Clearly there is something inconsistent with your setup parsing/writing. Please address. 

Thanks!
Joe

**Comment by profxj** [comment]
@jhennawi -- The incantation you used generated a psuedo-PypeIt file that was not intended to be used with `run_pypeit` only with `pypeit_seupt`.    As that is quite confusing, @kbwestfall and I agreed to shelve that concept (there were some other issues as well which I won't delve into).

I've also further modified `PypeItSetup` to generate only 2 files in the `setup_files/` folder:  a .sorted and a dummy .pypeit file that is clearly not for running.  No more .lst or .setups files.

As has always been the case, you must run `pypeit_setup` with `-c` to generate true PypeIt files.
That makes one file per setup placed in their own folders.

**Comment by jhennawi** [comment]
Okay thanks @profxj, can we just make -c the default, and abandon the psuedo PypeIt file concept. It sounds like that is what you have done anyway.

**Comment by profxj** [comment]
Yes, we should probably do that.
To be continued.

merging..

**Pull request merged**